x-pack/legacy/plugins/apm/public/components/app/ServiceMap/Cytoscape.tsx

```diff
@@ -52,30 +56,65 @@ function useCytoscape(options: cytoscape.CytoscapeOptions) {
   return [ref, cy] as [React.MutableRefObject<any>, cytoscape.Core | undefined];
 }
 
+function getLayoutOptions(
+  selectedRoots: string[],
+  height: number,
+  width: number
+): cytoscape.LayoutOptions {
+  return {
+    name: 'breadthfirst',
+    roots: selectedRoots,
+    fit: true,
+    padding: nodeHeight,
+    nodeDimensionsIncludeLabels: true,
+    animate: true,
+    animationEasing: animationOptions.easing,
+    animationDuration: animationOptions.duration,
+    // Rotate nodes from top -> bottom to display left -> right
+    // @ts-ignore
+    transform: (node: any, { x, y }: cytoscape.Position) => ({ x: y, y: -x }),
+    // swap width/height of boundingBox to compensation for the rotation
+    boundingBox: { x1: 0, y1: 0, w: height, h: width }
+  };
+}
+
+function selectRoots(elements: cytoscape.ElementDefinition[]): string[] {
+  const nodeData = elements
+    .filter(({ group }) => group === 'nodes')
+    .map(({ data }) => data) as cytoscape.NodeDataDefinition[];
+  const clientNodeIds = nodeData
+    .filter(data => isRumAgentName(data.agentName))
+    .map(({ id }) => id) as string[];
+  return clientNodeIds;
+}
+
 export function Cytoscape({
   children,
   elements,
   height,
+  width,
   serviceName,
   style
 }: CytoscapeProps) {
-  const [ref, cy] = useCytoscape({ ...cytoscapeOptions, elements });
+  const initialElements = elements.map(element => ({
+    ...element,
+    // prevents flash of unstyled elements
+    classes: [element.classes, 'invisible'].join(' ').trim()
+  }));
+  const [ref, cy] = useCytoscape({
+    ...cytoscapeOptions,
+    elements: initialElements
+  });
 
   // Add the height to the div style. The height is a separate prop because it
   // is required and can trigger rendering when changed.
   const divStyle = { ...style, height };
 
-  // Trigger a custom "data" event when data changes
-  useEffect(() => {
-    if (cy) {
-      cy.add(elements);
-      cy.trigger('data');
-    }
-  }, [cy, elements]);
-
-  // Set up cytoscape event handlers
-  useEffect(() => {
-    const dataHandler: cytoscape.EventHandler = event => {
+  const dataHandler = useCallback(
```

Comment:
The typing of this with the extra argument is weird now, so I think maybe there should be two functions, one that actually is an `EventHandler` (that you can type with `useCallback<cytoscape.EventHandler>` and the one that takes the extra argument that gets wrapped.

It's not a big thing, but I'd rather have more verbose code than confusing types.